runtime/compiler/net/ProtobufTypeConvert.hpp

```diff
@@ -0,0 +1,291 @@
+   extern uint64_t typeCount;
+   struct TypeID
+      {
+      uint64_t id;
+      TypeID()
+         {
+         id = typeCount++;
+         }
+      };
+
+   // used to unpack a tuple into variadic args
+   template <size_t... I> struct index_tuple { template <size_t N> using type = index_tuple<I..., N>; };
+   template <size_t N> struct index_tuple_gen { using type = typename index_tuple_gen<N - 1>::type::template type<N - 1>; };
+   template <> struct index_tuple_gen<0> { using type = index_tuple<>; };
+
+   // handles (de)serialization of "primitive" types, i.e. those that do not require more advanced pre/post processing
+   // this includes: int, bool, string, trivially copyable types, vector<T>, and tuple<T...> where T is primitive
+   template <typename T, typename = void> struct AnyPrimitive { };
+   template <> struct AnyPrimitive<uint32_t>
+      {
+      static inline uint32_t read(const Any *msg) { return msg->uint32_v(); }
+      static inline void write(Any* msg, const uint32_t val) { msg->set_uint32_v(val); }
+      static inline Any::TypeCase typeCase() { return Any::kUint32V; }
+      };
+   template <> struct AnyPrimitive<uint64_t>
+      {
+      static inline uint64_t read(const Any *msg) { return msg->uint64_v(); }
+      static inline void write(Any* msg, const uint64_t val) { msg->set_uint64_v(val); }
+      static inline Any::TypeCase typeCase() { return Any::kUint64V; }
+      };
+   template <> struct AnyPrimitive<int32_t>
+      {
+      static inline int32_t read(const Any *msg) { return msg->int32_v(); }
+      static inline void write(Any* msg, const int32_t val) { msg->set_int32_v(val); }
+      static inline Any::TypeCase typeCase() { return Any::kInt32V; }
+      };
+   template <> struct AnyPrimitive<int64_t>
+      {
+      static inline int64_t read(const Any *msg) { return msg->int64_v(); }
+      static inline void write(Any* msg, const int64_t val) { msg->set_int64_v(val); }
+      static inline Any::TypeCase typeCase() { return Any::kInt64V; }
+      };
+   template <> struct AnyPrimitive<bool>
+      {
+      static inline bool read(const Any *msg) { return msg->bool_v(); }
+      static inline void write(Any* msg, const bool val) { msg->set_bool_v(val); }
+      static inline Any::TypeCase typeCase() { return Any::kBoolV; }
+      };
+   template <> struct AnyPrimitive<const std::string>
+      {
+      static inline const std::string& read(const Any *msg) { return msg->bytes_v(); }
+      static inline void write(Any* msg, const std::string &val) { msg->set_bytes_v(val); }
+      static inline Any::TypeCase typeCase() { return Any::kBytesV; }
+      };
+   // trivially copyable
+   template <typename T> struct AnyPrimitive<T, typename std::enable_if<std::is_trivially_copyable<T>::value>::type>
+      {
+      static inline T read(const Any *msg)
+         {
+         char data[sizeof(T)];
+         const std::string &strVal = msg->bytes_v();
+         memcpy(data, &strVal[0], sizeof(T));
```

Comment:
`strVal.c_str()` would be better than assuming the data are contiguous starting at `&strVal[0]`.